### PR TITLE
Split define action

### PR DESCRIPTION
**Pull request by jaresty**
This pull request adds a command to split the window and follow in one command.

## Checklist

- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [ ] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [ ] I have not broken the cheatsheet

### CODE REVIEW

File: packages/cursorless-engine/src/actions/SimpleIdeCommandActions.ts
Comment (pokey) [inline packages/cursorless-engine/src/actions/SimpleIdeCommandActions.ts:109]: I'd remove this so that it restores the selection
```suggestion
```

File: data/fixtures/recorded/actions/splitDefineVest.yml
Comment (pokey) [inline data/fixtures/recorded/actions/splitDefineVest.yml:5]: This doesn't look right. Are you sure this wasn't recorded using your personal version from custom actions csv?

File: packages/common/src/types/TextEditor.ts
Comment (pokey) [inline packages/common/src/types/TextEditor.ts:203]: I would be tempted to just make this an option passed to `revealDefinition`. Slightly mixed feelings about it being on the editor, as it opens a new editor, but probably fine. Cc/ @AndreasArvidsson

File: packages/cursorless-vscode/src/ide/vscode/VscodeTextEditorImpl.ts
Comment (pokey) [inline packages/cursorless-vscode/src/ide/vscode/VscodeTextEditorImpl.ts:177]: Thinking on it, I think it would be better to have `followLinkAside`, because I haven't used `revealDefinition` in probably a year. Follow link is much more robust. Here is where that is implemented: https://github.com/cursorless-dev/cursorless/blob/main/packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts

You should be able to just do a case-matched find/replace on your PR to change the name, then there will be adding the options object and opening aside in VscodeOpenLink.ts

We could pass an options object indicating to open aside, as suggested in another of my comments

File: cursorless-talon/src/spoken_forms.json
Comment (pokey) [inline cursorless-talon/src/spoken_forms.json:20]: I'd be tempted to swap this: `"define split"`

**Review by pokey — COMMENTED** [review]
Nice job! This looks like it should be working. See my comments below, though

File: data/fixtures/recorded/actions/splitDefineVest.yml
Comment (AndreasArvidsson) [reply data/fixtures/recorded/actions/splitDefineVest.yml:5]: `version: 1`? I assume you just copied an existing file instead of recording a new one?

File: packages/common/src/types/TextEditor.ts
Comment (AndreasArvidsson) [reply packages/common/src/types/TextEditor.ts:203]: I'm fine with either way. You are revealing a definition of something that exists in a particular editor so I think it's fine that it's on the editor type.

File: cursorless-talon/src/spoken_forms.json
Comment (AndreasArvidsson) [reply cursorless-talon/src/spoken_forms.json:20]: Totally agree

File: packages/common/src/types/TextEditor.ts
Comment (pokey) [reply packages/common/src/types/TextEditor.ts:203]: Makes sense. WDYT about options object on existing function instead of new function. See also below re adding it to followLink instead of revealDefinition

Comment (AndreasArvidsson) [reply packages/common/src/types/TextEditor.ts:203]: As a said I'm fine with either adding an argument to existing function or adding a new function. With regards to the editor declaring capabilities it might be better with a separate function?

File: packages/cursorless-vscode/src/ide/vscode/VscodeTextEditorImpl.ts
Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/ide/vscode/VscodeTextEditorImpl.ts:177]: Same here; I don't use reveal definition at all anymore. Almost like we should deprecate that one?

File: data/fixtures/recorded/actions/splitDefineVest.yml
Comment (jaresty) [reply data/fixtures/recorded/actions/splitDefineVest.yml:5]: Yes, I copied the existing file because I wasn't able to figure out how to run the command from the debug editor-the test was passing so I was pretty sure the command was working, though.

Comment (pokey) [reply data/fixtures/recorded/actions/splitDefineVest.yml:5]: This doesn't actually test any of your code, though

File: packages/cursorless-vscode/src/ide/vscode/VscodeTextEditorImpl.ts
Comment (pokey) [reply packages/cursorless-vscode/src/ide/vscode/VscodeTextEditorImpl.ts:177]: Yeah I don't think there are any examples where it is better than follow link

File: data/fixtures/recorded/actions/splitDefineVest.yml
Comment (jaresty) [reply data/fixtures/recorded/actions/splitDefineVest.yml:5]: Ok, then I misunderstood how it works. I don't really understand how to do the local recording for test cases yet.

File: cursorless-talon/src/spoken_forms.json
Comment (jaresty) [reply cursorless-talon/src/spoken_forms.json:20]: Should it be "follow split"?

File: data/fixtures/recorded/actions/splitDefineVest.yml
Comment (pokey) [reply data/fixtures/recorded/actions/splitDefineVest.yml:5]: Assuming you found the docs for the test recorder?

File: cursorless-talon/src/spoken_forms.json
Comment (pokey) [reply cursorless-talon/src/spoken_forms.json:20]: Yeah

File: data/fixtures/recorded/actions/splitDefineVest.yml
Comment (jaresty) [reply data/fixtures/recorded/actions/splitDefineVest.yml:5]: I did. Will take a look when I have some more time. Thanks!

File: packages/cursorless-engine/src/actions/SimpleIdeCommandActions.ts
Comment (jaresty) [reply packages/cursorless-engine/src/actions/SimpleIdeCommandActions.ts]: Changed.

**Comment by jaresty** [comment]
I think I responded to all your comments. I wasn't sure if I should remove the define split commands. Everything seems to be working at the moment.

**Comment by AndreasArvidsson** [comment]
We probably gonna deprecate define in favor of follow. No sense adding another define then.

**Comment by jaresty** [comment]
I removed the define command. Only the follow command remains.

File: packages/cheatsheet/src/lib/sampleSpokenFormInfos/defaults.json
Comment (pokey) [inline packages/cheatsheet/src/lib/sampleSpokenFormInfos/defaults.json:170]: This is supposed to be generated automatically, but it's not easy to trigger today. For now I'd just manually change it to what the generated output would be
```suggestion
        {
          "id": "followLinkAside",
          "type": "action",
          "variations": [
            {
              "spokenForm": "follow split <target>",
              "description": "Follow link aside"
            }
          ]
        },
```

File: packages/cursorless-engine/src/actions/FollowLink.ts
Comment (pokey) [inline packages/cursorless-engine/src/actions/FollowLink.ts:29]: You should pass an options object here to indicate link should be opened aside, and then handle that [here](https://github.com/cursorless-dev/cursorless/blob/main/packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts#L66) by passing an options object there to indicate to open aside

**Review by pokey — COMMENTED** [review]
Looking pretty good; left a couple comments

File: packages/cursorless-engine/src/actions/FollowLink.ts
Comment (pokey) [inline packages/cursorless-engine/src/actions/FollowLink.ts:37]: Actually I think we should just move this whole `if` statement to somewhere in https://github.com/cursorless-dev/cursorless/blob/main/packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts? Feels pretty vscode-specific

WDYT @AndreasArvidsson

Comment (AndreasArvidsson) [reply packages/cursorless-engine/src/actions/FollowLink.ts:37]: Totally agree

Comment (pokey) [reply packages/cursorless-engine/src/actions/FollowLink.ts:37]: @jaresty I think makes sense to do that in this PR as it's a small change and you'll be touching that code already

Comment (jaresty) [reply packages/cursorless-engine/src/actions/FollowLink.ts:37]: Sounds good. I'll take a look soon.

Comment (jaresty) [reply packages/cursorless-engine/src/actions/FollowLink.ts:37]: I'm looking at this but I'm having a hard time understanding what the change needs to be. It seems that the VSCode follow link method currently only handles URIs. I'm not quite sure how to make the change without doing a significant restructure here, which makes me think that I'm misunderstanding something about the intention or the request. Can you give me a little bit more specific details about what you're hoping to see change? I also don't mind if you want to make the change yourself (in case it's easier to do it than to explain).

Comment (jaresty) [reply packages/cursorless-engine/src/actions/FollowLink.ts]: I think I figured it out. One thing that I'm still unsure about is that in the previous version I needed to pass the targets to the execute action, but now it doesn't seem to be necessary. The functionality does seem to work however.

**Comment by jaresty** [comment]
I think I addressed all the comments. Anything I missed?

File: data/fixtures/recorded/actions/followSplitVest.yml
Comment (AndreasArvidsson) [inline data/fixtures/recorded/actions/followSplitVest.yml:7]: This doesn't test the new implementation

File: packages/cursorless-engine/src/actions/FollowLink.ts
Comment (AndreasArvidsson) [inline packages/cursorless-engine/src/actions/FollowLink.ts:16]: I would probably just do `private options : { openInSplit: boolean }`.

File: data/fixtures/recorded/actions/followSplitVest.yml
Comment (jaresty) [reply data/fixtures/recorded/actions/followSplitVest.yml]: Shoot. Apparently it was picking up my custom action instead. I'm having some trouble getting it to recognize this new action now. Do you have any tips to debug if it's not seeming to recognize the action at all?

Comment (jaresty) [reply data/fixtures/recorded/actions/followSplitVest.yml]: I forgot that I needed to symlink the cursor list settings in place of my own settings. Updated with the new test case.

File: packages/cursorless-engine/src/actions/FollowLink.ts
Comment (AndreasArvidsson) [inline packages/cursorless-engine/src/actions/FollowLink.ts:19]: When you use private on a constructor argument you don't need to assign it yourself; that is done automatically
Just remove lines `openInSplit = false;` and `this.options = options;`

File: packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts
Comment (AndreasArvidsson) [inline packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts:22]: const instead of let

File: packages/cursorless-engine/src/actions/FollowLink.ts
Comment (AndreasArvidsson) [inline packages/cursorless-engine/src/actions/FollowLink.ts:27]: This return value is not used anymore. Should probably be removed from the openLink type.

File: packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts
Comment (AndreasArvidsson) [inline packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts:25]: Also you still need a return here. Otherwise you will run open link a few lines below on undefined.

**Review by AndreasArvidsson — CHANGES_REQUESTED** [review]

File: packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts
Comment (AndreasArvidsson) [inline packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts:25]: I don't think this will actually work. You are executing reveal definition on the current selection always and not on the actual target. I would probably still use our own `actions.executeCommand`

**Review by AndreasArvidsson — CHANGES_REQUESTED** [review]

File: packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts
Comment (jaresty) [reply packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts:25]: I think you're right. I didn't change this though-don't really understand what you mean by 'use our own actions.executeCommand'-can you help me understand?

Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts:25]: The one you removed in this pull request

https://github.com/cursorless-dev/cursorless/blob/766864d9fa979124283b236bbb4738204a13fe45/packages/cursorless-
engine/src/actions/FollowLink.ts#L27-L31

But since we are already setting selection for the `editor.action.openLink` command it's probably best to merge these behaviors

Comment (jaresty) [reply packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts:25]: oh okay I see it now

Comment (jaresty) [reply packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts]: I wasn't sure exactly what you meant but I gave it a shot; let me know how it looks. Thanks!

Comment (AndreasArvidsson) [reply packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts]: You still need the if statement to check if we actually have found a link. The set selection code could be extracted to ear separate function that is used in both cases.
Also I would remove the return boolean type and just have void.

Comment (jaresty) [reply packages/cursorless-vscode/src/ide/vscode/VscodeOpenLink.ts:29]: 1. Extracted the set selection code
2. the return type is harder-that type is defined in the interface. Are you sure I should remove it?

**Review by pokey — APPROVED** [review]
I made some tweaks in f5d58fe2a9c0ea158551b93a704a3dc6d53d420b. I'm happy, but I believe will require approval from @AndreasArvidsson before Github will let it merge as he's requested changes

**Comment by jaresty** [comment]
Thanks for cleaning it up! @pokey

**Comment by AndreasArvidsson** [comment]
Nice exactly what I wanted! :)

**Comment by pokey** [comment]
> Nice exactly what I wanted! :)

I didn't actually read any of your comments, so it's a good sign that we independently arrived at the same conclusion 😄

**Comment by AndreasArvidsson** [comment]
> > Nice exactly what I wanted! :)
> 
> I didn't actually read any of your comments, so it's a good sign that we independently arrived at the same conclusion 😄

Heh don't know if I should be hurt or impressed by that :D

**Comment by pokey** [comment]
well they had all been marked as resolved so I assumed they'd already been addressed 😅

**Comment by jaresty** [comment]
Oops, sorry. I was using them kind of like a todo list-should I have left them open?  Appreciate all the guidance-thanks for helping me get this done 😀

**Comment by pokey** [comment]
> Oops, sorry. I was using them kind of like a todo list-should I have left them open?  Appreciate all the guidance-thanks for helping me get this done 😀

That's ok. If you're confident you've resolved them, feel free to mark them resolved. If you have any doubt, though, best to leave them open and let us have a look to see if they've been resolved

**Pull request merged**
